src/hazelcore/semantics/UHPat.re

```diff
@@ -153,6 +158,51 @@ and make_inconsistent_operand =
     (Parenthesized(set_p), u_gen);
   };
 
+let patterns_of_type =
+    (u_gen: MetaVarGen.t, ty: HTyp.t): (list(t), MetaVarGen.t) =>
+  OpSeq.(
+    switch (ty) {
+    | HTyp.Hole
+    | Unit
+    | Arrow(_, _) => ([], u_gen)
+    | Num => ([numlit(0), wild()] |> List.map(wrap), u_gen)
+    | Bool => ([true, false] |> List.map(b => wrap(boollit(b))), u_gen)
+    | Prod(_, _) =>
+      let prod_element_count = ty |> HTyp.get_prod_elements |> List.length;
+      let patternSkel =
+        ListUtil.range(prod_element_count)
+        |> List.map(i => Skel.Placeholder(i))
+        |> make_tuple(NotInHole);
+      let opseq_of_holes = u_gen => {
+        let (firstHole, u_gen) = u_gen |> new_EmptyHole;
+        let (holes, u_gen) = new_EmptyHoles(prod_element_count - 1, u_gen);
+        let opseq =
+          OpSeq(
+            patternSkel,
+            Seq.mk(firstHole, holes |> List.map(hole => (Comma, hole))),
+          );
+        (opseq, u_gen);
+      };
+
+      let (opseq1, u_gen) = opseq_of_holes(u_gen);
+      let (opseq2, u_gen) = opseq_of_holes(u_gen);
+      ([opseq1, opseq2], u_gen);
```

Comment:
why does this return two patterns? there should only be one I think...

Reply:
I thought about this as well for a little bit, but what use is a `case` expression with a single pattern? For safe code, you'll always want at least 2.

Reply:
that's true when you are at the top-level of a case expression, but we can have this mechanism work within sub-patterns too (e.g. if you had a list of tuples, the first invocation would create the nil and cons, but then you might want to invoke it again in head position of the cons pattern)

there is only one pattern form for tuples -- the tuple pattern

Reply:
Okay, got it. I'll change it to one.